File: tools/tourism/meet.py
"""The human layer: /meet, built from the dataset.

    python3 tools/tourism/build.py meet

Every country in this set was written up through the same twenty-seven
categories. Twelve of them are about people — who lives here, what is cooked,
what is made by hand, what is kept, what a city is like on a Tuesday — and until
now they sat in a grid of twenty-seven tiles where the twelve looked exactly like
the fifteen about scenery.

This page is the argument that they are not the same thing. It has two axes:

    one country, seven doors     — Cameroon: people, culture, food, craft,
                                   cities, heritage, responsibility
    one door, twenty-two countries — food, asked of the whole continent at once

The second is the point. A tourism site can always find one photograph of one
market. What it cannot do is ask the same question of twenty-two countries and
print twenty-two different answers, each naming a particular people, a
particular dish, a particular workshop in a particular town. The uniform
taxonomy makes that possible and nothing else on the site uses it that way.

Nothing here is invented. Every line behind every door is a caption and a
description already written for that country. Guides come from people.json and
operator notes from voices.json, both of which are empty and stay empty until
somebody has asked a real person — the components render nothing rather than
render a plausible stranger.

What ships:

    meet.html        the page and the spine
    data/meet.json   seven strands x twenty-two countries, fetched once
"""

import html as html_mod
import json
import os

from . import plate
from .model import (ROOT, load_operators, load_people, load_regions,
                    load_strands, load_voices)

PAGE = os.path.join(ROOT, "meet.html")
DATA = os.path.join(ROOT, "data", "meet.json")


def esc(v):
    return html_mod.escape(str(v if v is not None else ""), quote=True)


# ---- the payload -----------------------------------------------------------------


def strands_payload(countries, taxonomy):
    """Seven doors x twenty-two countries, in one file.

    It is the same source as data/atlas/<slug>.json, cut the other way. The
    atlas needs one country at a time and fetches one file; this page needs one
    question across the whole continent and would otherwise fetch twenty-two.
    Both are generated from tourism/countries/*.json in the same build, so
    neither can drift from the other.
    """
    strands = load_strands()
    ops = load_operators()
    titles = {c["id"]: c["title"] for c in taxonomy.categories}
    live = [c for c in countries if c.published]

    out = {"strands": [], "countries": {}, "answers": {}}
    for key, s in strands.items():
        out["strands"].append({"key": key, "title": s.get("title") or key,
                               "asks": s.get("asks") or "",
                               "line": s.get("line") or "",
                               "categories": s.get("categories") or []})

    for c in live:
        op = ops.get(c.operator_key)
        out["countries"][c.slug] = {
            "name": c.name, "adjective": c.adjective, "region": c.region,
            "tagline": c.tagline, "url": c.url,
            "window": c.window, "windowAlt": c.window_alt,
            "operator": ({"key": c.operator_key, "name": op.name, "base": op.base,
                          "since": op.since, "line": op.line, "url": op.url}
                         if op else None),
        }

    for key, s in strands.items():
        rows = {}
        for c in live:
            answers = []
            for cat in s.get("categories") or []:
                e = c.entry(cat)
                if not e or not e.caption:
                    continue
                img = e.image or {}
                answers.append({
                    "id": cat, "group": titles.get(cat, cat),
                    "title": e.caption, "text": e.description or "",
                    "image": ({"url": img.get("imageUrl"),
                               "alt": img.get("alt") or e.description,
                               "credit": img.get("photographer"),
                               "provider": img.get("provider")}
                              if img.get("imageUrl") else None),
                })
            if answers:
                rows[c.slug] = answers
        out["answers"][key] = rows
    return out


def people_payload():
    """Real people, if any have been recorded. Usually none, deliberately."""
    people = load_people()
    ops = load_operators()
    out = []
    for key, p in people.items():
        op = ops.get(p.operator_key)
        out.append({
            "key": key, "name": p.name, "role": p.role, "country": p.country,
            "base": p.base, "languages": p.languages, "since": p.since,
            "speciality": p.speciality, "line": p.line,
            "photo": p.photo, "photoAlt": p.photo_alt,
            "operator": ({"name": op.name, "url": op.url} if op else None),
        })
    return out


def voices_payload():
    """What operators have actually said. Empty until somebody says something."""
    ops = load_operators()
    out = []
    for v in load_voices():
        op = ops.get(v.operator_key)
        if not op:
            continue
        out.append({"country": v.country, "asked": v.asked, "said": v.said,
                    "by": v.by, "operator": {"name": op.name, "base": op.base,
                                             "url": op.url}})
    return out


# ---- the page --------------------------------------------------------------------


def door_list(data):
    """The seven doors, written into the page so they exist without script."""
    out = []
    for i, s in enumerate(data["strands"]):
        n = len(data["answers"].get(s["key"]) or {})
        out.append(
            '        <li>\n'
            '          <button class="mt-door" type="button" data-strand="%s" '
            'aria-pressed="%s">\n'
            '            <span class="mt-door-no">%02d</span>\n'
            '            <span class="mt-door-body"><b>%s</b>'
            '<span class="mt-door-asks">%s</span></span>\n'
            '            <span class="mt-door-n">%d</span>\n'
            '          </button>\n        </li>'
            % (esc(s["key"]), "true" if i == 0 else "false", i + 1,
               esc(s["title"]), esc(s["asks"]), n))
    return "\n".join(out)


def country_strip(data):
    """Every country, in the order the continent is read, as real links.

    With no script these are links to the country's own page, so the page is a
    working index of twenty-two countries before anything runs.
    """
    regions = load_regions()
    order = {}
    for i, (key, reg) in enumerate(regions.items()):
        for inc in reg.includes:
            order[inc] = i
    rows = sorted(data["countries"].items(),
                  key=lambda kv: (order.get(kv[1]["region"], 99), kv[1]["name"]))
    return "\n            ".join(
        '<a class="mt-flag" href="%s" data-country="%s">%s</a>'
        % (esc(c["url"]), esc(slug), esc(c["name"])) for slug, c in rows)


def opening(data):
    """The first thing behind the first door, so the page is never blank.

    Rendered from the payload at build time rather than drawn by script: the
    continent-wide answer to the first question is the whole argument of the
    page, and it should be readable with JavaScript switched off.
    """
    first = data["strands"][0]
    rows = data["answers"].get(first["key"]) or {}
    out = []
    regions = load_regions()
    order = {}
    for i, (key, reg) in enumerate(regions.items()):
        for inc in reg.includes:
            order[inc] = i
    for slug in sorted(rows, key=lambda s: (order.get(data["countries"][s]["region"], 99),
                                            data["countries"][s]["name"])):
        c = data["countries"][slug]
        a = rows[slug][0]
        out.append(
            '      <article class="mt-answer">\n'
            '        <p class="mt-answer-where"><a href="%s">%s</a>'
            '<span>%s</span></p>\n'
            '        <h3 class="mt-answer-title">%s</h3>\n'
            '        <p class="mt-answer-text">%s</p>\n'
            '      </article>'
            % (esc(c["url"]), esc(c["name"]), esc(c["region"]),
               esc(a["title"]), esc(a["text"])))
    return "\n".join(out)


def render(countries, taxonomy):
    data = strands_payload(countries, taxonomy)
    if not data["countries"]:
        raise IOError("no published countries — nothing to meet")
    people = people_payload()
    return TEMPLATE % {
        "events": plate.events_block(),
        "explore": plate.explore_block(),
        "foot": plate.colophon_foot("/meet"),
        "og": plate.open_graph('Meet Africa — Afrinkong', 'Seven questions, asked of twenty-two countries. The same question changes its answer at every border.', '/meet'),
        "doors": door_list(data),
        "strip": country_strip(data),
        "opening": opening(data),
        "first": esc(data["strands"][0]["title"]),
        "firstAsks": esc(data["strands"][0]["asks"]),
        "firstLine": esc(data["strands"][0]["line"]),
        "n": len(data["countries"]),
        "doorCount": len(data["strands"]),
        "answers": sum(len(v) for v in data["answers"].values()),
        # People are inlined because there are none: an empty array is twelve
        # bytes and a request for twelve bytes is a request too many. The day
        # somebody is added this stays true until the file is large enough to
        # be worth splitting, which is a long way off.
        "people": json.dumps(people, separators=(",", ":")),
        "voices": json.dumps(voices_payload(), separators=(",", ":")),
    }


def run(countries, taxonomy, log=print):
    data = strands_payload(countries, taxonomy)
    folder = os.path.dirname(DATA)
    if not os.path.isdir(folder):
        os.makedirs(folder)
    with open(DATA, "w") as fh:
        json.dump(data, fh, separators=(",", ":"), sort_keys=True)
        fh.write("\n")
    html = render(countries, taxonomy)
    with open(PAGE, "w") as fh:
        fh.write(html)
    answers = sum(len(v) for v in data["answers"].values())
    people = len(load_people())
    log("meet: %s (%.1f KB) + %s (%.1f KB), %d doors x %d countries = %d answers"
        % (os.path.relpath(PAGE, ROOT), len(html) / 1024.0,
           os.path.relpath(DATA, ROOT), os.path.getsize(DATA) / 1024.0,
           len(data["strands"]), len(data["countries"]), answers))
    log("      %d guide profile(s), %d operator note(s) — both empty until a real "
        "person has been asked" % (people, len(load_voices())))
    return PAGE


TEMPLATE = """<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Meet Africa &mdash; Afrinkong</title>
<meta name="description" content="Seven questions, asked of twenty-two countries. Who lives here, what is cooked, what is made by hand, what is kept — and the same question changes its answer at every border.">
%(og)s
<link rel="stylesheet" href="/styles/afrinkong.css">
<link rel="stylesheet" href="/styles/meet.css">
</head>
<body>
<a class="af-skip" href="#doors">Skip to the questions</a>
<header class="mt-mast">
  <a class="mt-mark" href="/"><i>Afrinkong</i><b>Meet Africa</b></a>
  <nav class="mt-routes" aria-label="Primary">
    <a href="/atlas">The Atlas</a>
    <a href="/journey">Build a journey</a>
    <a href="/places">Every place</a>
    <a href="/stories">Stories</a>
  </nav>
  <a class="af-btn af-btn--quiet" href="/enquire">Talk to us<i>&rarr;</i></a>
</header>

<main class="mt" id="mt" data-mode="strand">

  <section class="mt-open">
    <span class="af-stamp">The human layer</span>
    <h1 class="mt-h1">Africa is not a backdrop.</h1>
    <p class="mt-lede">Every country here was written up through the same
      twenty-seven headings, and twelve of them are about people. So the same
      question can be put to all %(n)d at once &mdash; and it comes back %(n)d
      different ways, which is the part a photograph of a sunset cannot tell you.
      %(doorCount)d questions, %(answers)d answers, none of them about
      &ldquo;African culture&rdquo; in general.</p>
  </section>

  <div class="mt-grid">
    <nav class="mt-doors" id="doors" aria-label="The seven questions">
      <ol>
%(doors)s
      </ol>
      <p class="mt-doors-note">Pick a country to put all seven to one place.</p>
      <div class="mt-strip" role="group" aria-label="Countries">
            %(strip)s
      </div>
    </nav>

    <section class="mt-stage" id="stage" aria-live="polite">
      <header class="mt-stage-head" id="stage-head">
        <span class="af-stamp" id="mt-eyebrow">%(n)d countries</span>
        <h2 class="mt-h2" id="mt-asks">%(firstAsks)s</h2>
        <p class="mt-stage-line" id="mt-line">%(firstLine)s</p>
      </header>
      <div class="mt-answers" id="mt-answers">
%(opening)s
      </div>
    </section>
  </div>

  <noscript>
    <p class="mt-nojs">The seven questions above are shown one at a time, which
      needs JavaScript &mdash; without it you are reading the first one,
      answered by all %(n)d countries. Every country name is a link to its own
      page, where all twenty-seven headings are written out in full.</p>
  </noscript>
</main>
%(foot)s

<script type="application/json" id="mt-people">%(people)s</script>
<script type="application/json" id="mt-voices">%(voices)s</script>
%(events)s
%(explore)s
<script src="/scripts/window.js" defer></script>
<script src="/scripts/meet.js" defer></script>
</body>
</html>
"""

File: scripts/meet.js
/* The human layer — two ways of asking the same seven questions.
 *
 * @product: live | @gate: none | @surface: /meet
 * ---------------------------------------------------------------------------
 *   one door,  twenty-two countries   #/food
 *   one country, seven doors          #/cameroon
 *   one answer                        #/food/cameroon
 *
 * The first mode is the argument of the page: the same question put to the
 * whole continent, coming back twenty-two different ways. The second is the
 * one anybody planning a trip actually wants. They are the same data read
 * along two axes, so there is one payload and one render.
 *
 * Nothing here writes a sentence. Every line on screen is a caption or a
 * description already written for that country, an operator's own words, or
 * this file's own labels — and the guide component renders nothing at all
 * while people.json is empty, which is the point of it.
 */
(function () {
  'use strict';

  var root = document.getElementById('mt');
  var doors = document.getElementById('doors');
  var stage = document.getElementById('stage');
  var box = document.getElementById('mt-answers');
  var eyebrow = document.getElementById('mt-eyebrow');
  var asks = document.getElementById('mt-asks');
  var line = document.getElementById('mt-line');
  if (!root || !box) return;

  var PEOPLE = readJSON('mt-people', []);
  var VOICES = readJSON('mt-voices', []);
  var reduced = matchMedia('(prefers-reduced-motion: reduce)');
  var track = function (name, props) {
    if (window.AfrinkongEvents) window.AfrinkongEvents.track(name, props);
  };

  var D = null;                 /* data/meet.json, fetched once */
  var state = {strand: null, country: null};
  var order = [];               /* countries, in the order the continent reads */

  function readJSON(id, fallback) {
    var el = document.getElementById(id);
    if (!el) return fallback;
    try { return JSON.parse(el.textContent); } catch (e) { return fallback; }
  }
  function esc(s) {
    return String(s == null ? '' : s).replace(/[&<>"]/g, function (c) {
      return {'&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;'}[c];
    });
  }
  function strandBy(key) {
    for (var i = 0; i < D.strands.length; i++) {
      if (D.strands[i].key === key) return D.strands[i];
    }
    return D.strands[0];
  }

  /* ---- the two views ------------------------------------------------------ */

  /* One question, every country. Ordered geographically rather than
     alphabetically, so reading down the page is a sweep across the continent
     and the differences land as differences between neighbours. */
  function paintStrand() {
    var s = strandBy(state.strand);
    var rows = D.answers[s.key] || {};
    var slugs = order.filter(function (x) { return rows[x]; });
    eyebrow.textContent = slugs.length + ' countries · ' + s.title;
    asks.textContent = s.asks;
    line.textContent = s.line;
    box.className = 'mt-answers';
    box.innerHTML = slugs.map(function (slug) {
      var c = D.countries[slug];
      var a = rows[slug][0];
      return '<article class="mt-answer">'
        + '<p class="mt-answer-where"><button type="button" data-open="' + esc(slug)
        + '">' + esc(c.name) + '</button><span>' + esc(c.region) + '</span></p>'
        + '<h3 class="mt-answer-title">' + esc(a.title) + '</h3>'
        + '<p class="mt-answer-text">' + esc(a.text) + '</p>'
        + (rows[slug].length > 1
           ? '<p class="mt-answer-more">' + esc(rows[slug][1].title) + '</p>' : '')
        + '</article>';
    }).join('');
  }

  /* One country, all seven doors. The door keeps its universal word and shows
     that country's particular answer under it, which is what stops the page
     reading as seven generic cards. */
  function paintCountry() {
    var c = D.countries[state.country];
    if (!c) { paintStrand(); return; }
    eyebrow.textContent = c.region + ' · ' + D.strands.length + ' questions';
    asks.textContent = c.name;
    line.textContent = c.tagline;
    box.className = 'mt-answers mt-answers--country';
    var rooms = D.strands.map(function (s, i) {
      var rows = (D.answers[s.key] || {})[state.country];
      if (!rows || !rows.length) {
        return '<article class="mt-room mt-room--empty">'
          + '<p class="mt-room-eye">' + esc(s.title) + '</p>'
          + '<p class="mt-room-none">Not written up for ' + esc(c.name)
          + ' yet.</p></article>';
      }
      var open = state.strand === s.key;
      return '<article class="mt-room"' + (open ? ' data-open' : '') + '>'
        + '<button class="mt-room-hit" type="button" data-room="' + esc(s.key) + '"'
        + ' aria-expanded="' + (open ? 'true' : 'false') + '">'
        + '<span class="mt-room-no">' + (i < 9 ? '0' : '') + (i + 1) + '</span>'
        + '<span class="mt-room-body"><span class="mt-room-eye">' + esc(s.title)
        + '</span><b>' + esc(rows[0].title) + '</b></span>'
        + '<span class="mt-room-go" aria-hidden="true">&rarr;</span></button>'
        + '<div class="mt-room-in">'
        + rows.map(function (a, n) {
            return (n ? '<h4 class="mt-room-sub">' + esc(a.title) + '</h4>' : '')
              + '<p class="mt-room-text">' + esc(a.text) + '</p>'
              + '<p class="mt-room-cat">' + esc(a.group) + '</p>';
          }).join('')
        + '</div></article>';
    }).join('');

    box.innerHTML = '<div class="mt-country">'
      + '<div class="mt-country-shape" id="mt-shape"></div>'
      + '<div class="mt-rooms">' + rooms + '</div>'
      + '</div>'
      + voiceBlock(c) + peopleBlock(c)
      + '<div class="mt-acts">'
      + '<a class="af-btn af-btn--solid" href="/journey#/j/' + esc(state.country)
      + '/">Build a journey here<i>&rarr;</i></a>'
      + '<a class="af-btn af-btn--quiet" href="/portrait/' + esc(state.country)
      + '">Read the portrait</a>'
      + '<a class="af-btn af-btn--quiet" href="/atlas#/' + esc(state.country)
      + '">Find it on the map</a>'
      + '<a class="af-btn af-btn--quiet" href="' + esc(c.url) + '">All twenty-seven</a>'
      + '</div>';
    drawShape(state.country);
  }

  /* The silhouette is fetched only when a country is opened, and only once —
     twenty-nine kilobytes of outlines is not worth carrying to somebody who
     came to read about food. */
  function drawShape(slug) {
    var host = document.getElementById('mt-shape');
    if (!host) return;
    window.AfrinkongWindow.shapes().then(function (all) {
      var s = all[slug], c = D.countries[slug];
      if (!s || !c || !host.isConnected) return;
      host.toggleAttribute('data-photo', !!c.window);
      host.innerHTML = window.AfrinkongWindow.svg(s, {
        image: c.window, alt: c.windowAlt, name: c.name, ident: 'mtw',
        classes: 'af-window-svg'})
        + '<span class="mt-country-cap">' + esc(c.name) + '</span>';
    });
  }

  /* ---- the people components --------------------------------------------- */

  /* Local Voice. Where an operator has written something down it is printed in
     their name. Where they have not, the block still appears — but it says who
     the company is and what they run, in their own sentence, and anything the
     site itself knows is attributed to the site. Nothing is put in anybody's
     mouth. */
  function voiceBlock(c) {
    var said = VOICES.filter(function (v) { return v.country === state.country; })[0];
    if (said) {
      return '<section class="mt-voice"><span class="af-stamp">Local voice</span>'
        + (said.asked ? '<p class="mt-voice-q">' + esc(said.asked) + '</p>' : '')
        + '<blockquote class="mt-voice-said">&ldquo;' + esc(said.said) + '&rdquo;</blockquote>'
        + '<p class="mt-voice-by">' + esc(said.by ? said.by + ', ' : '')
        + esc(said.operator.name) + ' &mdash; ' + esc(said.operator.base) + '</p>'
        + '</section>';
    }
    /* Where there is no operator there is no quote, so nothing is rendered.
       The block that stood here printed a paragraph announcing the absence —
       "there is nobody here to quote, and we will not write a note on behalf
       of a company we have not named" — which is a section whose entire
       content is an apology for not being a section. Not writing a fake quote
       is right; devoting a panel to saying so is not. */
    if (!c.operator) return '';
    return '<section class="mt-voice"><span class="af-stamp">Local voice</span>'
      + '<blockquote class="mt-voice-said">&ldquo;' + esc(c.operator.line)
      + '&rdquo;</blockquote>'
      + '<p class="mt-voice-by">' + esc(c.operator.name) + ' &mdash; '
      + esc(c.operator.base) + (c.operator.since ? ', since ' + esc(c.operator.since) : '')
      + '</p>'
      + '<p class="mt-voice-note">That is the company’s own description of what '
      + 'it runs. Anything else on this page is Afrinkong’s writing, not theirs.</p>'
      + '</section>';
  }

  /* Guides. people.json is empty and this renders nothing while it is — a
     fabricated guide is worse than an absent one, because a visitor cannot tell
     the difference until they arrive. */
  function peopleBlock(c) {
    var here = PEOPLE.filter(function (p) { return p.country === state.country; });
    if (!here.length) return '';
    return '<section class="mt-people"><span class="af-stamp">Who might take you</span>'
      + '<ul class="mt-people-list">' + here.map(function (p) {
        return '<li class="mt-person">'
          + (p.photo ? '<img src="' + esc(p.photo) + '" alt="' + esc(p.photoAlt)
             + '" loading="lazy" width="240" height="300">' : '')
          + '<div><b>' + esc(p.name) + '</b>'
          + '<span class="mt-person-role">' + esc(p.role)
          + (p.base ? ' &middot; ' + esc(p.base) : '') + '</span>'
          + (p.line ? '<p>' + esc(p.line) + '</p>' : '')
          + '<dl class="mt-person-facts">'
          + (p.languages && p.languages.length
             ? '<dt>Languages</dt><dd>' + esc(p.languages.join(' · ')) + '</dd>' : '')
          + (p.since ? '<dt>Guiding since</dt><dd>' + esc(p.since) + '</dd>' : '')
          + (p.speciality ? '<dt>Speciality</dt><dd>' + esc(p.speciality) + '</dd>' : '')
          + '</dl></div></li>';
      }).join('') + '</ul></section>';
  }

  /* ---- chrome ------------------------------------------------------------ */

  function paintDoors() {
    [].forEach.call(doors.querySelectorAll('[data-strand]'), function (b) {
      b.setAttribute('aria-pressed', String(b.dataset.strand === state.strand
        && !state.country));
    });
    [].forEach.call(doors.querySelectorAll('[data-country]'), function (a) {
      a.toggleAttribute('data-on', a.dataset.country === state.country);
    });
    root.dataset.mode = state.country ? 'country' : 'strand';
  }

  function render() {
    paintDoors();
    if (state.country) paintCountry(); else paintStrand();
  }

  function go(next, push) {
    state = {strand: next.strand || state.strand, country: next.country || null};
    if (!D.answers[state.strand]) state.strand = D.strands[0].key;
    var h = '#/' + (state.country ? state.country + (next.strand ? '/' + state.strand : '')
                                  : state.strand);
    if (push !== false && h !== location.hash) history.pushState(null, '', h);
    if (state.country) track('meet_country_opened', {country: state.country});
    else track('meet_strand_opened', {strand: state.strand});
    render();
    stage.scrollIntoView({behavior: reduced.matches ? 'auto' : 'smooth', block: 'start'});
  }

  function fromHash(push) {
    var parts = (location.hash || '').replace(/^#\/?/, '').split('/').filter(Boolean);
    var strand = null, country = null;
    parts.forEach(function (p) {
      if (D.answers[p]) strand = p;
      else if (D.countries[p]) country = p;
    });
    state.strand = strand || D.strands[0].key;
    state.country = country;
    render();
    if (push) history.replaceState(null, '', '#/' + (country || state.strand));
  }

  document.addEventListener('click', function (e) {
    var t = e.target.closest ? e.target.closest(
      '[data-strand],[data-country],[data-open],[data-room]') : null;
    if (!t) return;
    if (t.hasAttribute('data-strand')) { go({strand: t.dataset.strand, country: null}); }
    else if (t.hasAttribute('data-country')) { e.preventDefault(); go({country: t.dataset.country}); }
    else if (t.hasAttribute('data-open')) { go({country: t.dataset.open}); }
    else if (t.hasAttribute('data-room')) {
      /* Inside a country the doors open in place rather than navigating: you
         are already in the room, and being thrown back to the top of the page
         to read the next paragraph is not opening a door. */
      var room = t.closest('.mt-room');
      var open = room.hasAttribute('data-open');
      [].forEach.call(box.querySelectorAll('.mt-room[data-open]'), function (r) {
        r.removeAttribute('data-open');
        r.querySelector('[aria-expanded]').setAttribute('aria-expanded', 'false');
      });
      if (!open) {
        room.setAttribute('data-open', '');
        t.setAttribute('aria-expanded', 'true');
        state.strand = t.dataset.room;
        history.replaceState(null, '', '#/' + state.country + '/' + state.strand);
      }
    }
  });

  /* Escape climbs back out of a country to the continent-wide view, which is
     the only place there is to go. */
  document.addEventListener('keydown', function (e) {
    if (e.key === 'Escape' && state.country) go({country: null, strand: state.strand});
  });

  window.addEventListener('popstate', function () { fromHash(false); });

  fetch('/data/meet.json').then(function (r) { return r.json(); })
    .then(function (j) {
      D = j;
      /* Geographic order, taken from the strip the page already printed, so
         the two agree and neither has to carry a second copy of it. */
      order = [].map.call(doors.querySelectorAll('[data-country]'),
        function (a) { return a.dataset.country; });
      fromHash(true);
    })
    .catch(function () {
      /* The first question, answered by everybody, is already on the page:
         if the payload never arrives the visitor still has that rather than
         an empty frame. */
      root.setAttribute('data-offline', '');
    });
})();
